src/gfx.cpp

```diff
@@ -154,6 +154,127 @@ void GfxFillRect(int left, int top, int right, int bottom, int colour, FillRectM
 	}
 }
 
+/**
+ * Fill a polygon with colour.
+ * For rectangles the GfxFillRect function will be faster.
+ * @pre dpi->zoom == ZOOM_LVL_NORMAL
+ * @param shape List of points on the polygon, must contain at least 3 points.
+ * @param colour A 8 bit palette index (FILLRECT_OPAQUE and FILLRECT_CHECKER) or a recolour spritenumber (FILLRECT_RECOLOUR).
+ * @param mode
+ *         FILLRECT_OPAQUE:   Fill the polygon with the specified colour
+ *         FILLRECT_CHECKER:  Like FILLRECT_OPAQUE, but only draw every second pixel (used to grey out things)
+ *         FILLRECT_RECOLOUR:  Apply a recolour sprite to every pixel in the rectangle currently on screen
+ */
+void GfxFillPolygon(const std::vector<Point> &shape, int colour, FillRectMode mode)
+{
+	typedef std::pair<Point, Point> Segment;
+	std::vector<Segment> segments;
+	Blitter *blitter = BlitterFactory::GetCurrentBlitter();
+	const DrawPixelInfo *dpi = _cur_dpi;
+
+	if (dpi->zoom != ZOOM_LVL_NORMAL) return;
+	assert(shape.size() >= 3);
+
+	{
```

Comment:
why this block?

Reply:
Just to keep the `prev` and `first` variables from sticking around in scope. Honestly it should probably be extracted to a function on its own.